docs/rfcs/0003_template_fields_mvp.md

```diff
@@ -0,0 +1,406 @@
+
+**Annotate Template**
+
+Tell us what your columns mean to enable automatic validation
+
+1. Analyst: Browse keys or canonical names and assigns the relevant one to each field in the template
+2. Analyst: Submit updated template
+3. API: Automatically generate an initial Extract Config
+   which simply maps columns in the source file to the keys defined in the template
+
+**Add Data Wrangling Functions**
+
+1. Analyst: Download Extract Config for a template
+1. Analyst: Edit it to include additional wrangling logic
+1. Upload Extract Config for the template
+
+**Submit Data**
+
+1. Data Submitter: Upload data file claiming to conform to a template
+2. API: Perform basic single file validation
+3. Data Submitter: Receive immediate validation feedback on data file (e.g. missin required column "Gender")
+4. Data Submitter: Fix any errors in file, repeat steps 1-3
+
+**Review and Validate Submission**
+
+1. Analyst: Create data review and add data file to the review
+2. Analyst: Click button to validate data files in review
+3. API: Clean and extract data files using the Extract Config for the
+   file
+4. API: Run validation on cleaned data files
+5. API: Generate validation report
+6. Analyst: View validation report on Data Tracker and take actions
+
+The `Review and Validation Submission` features is actually already implemented
+in Data Tracker and Study Creator API, thus this approach is very feasible for an MVP
+solution.
+
+## API Details
+
+```python
+
+class TemplateVersion:
+
+    @classmethod
+    def validate_field_definitions(cls):
+        """
+        Use json schema to validate the field_definitions structure and content
+        """
+        pass
+
+    def field_definitions_to_template(self, output_filepath=None):
+        """
+        Generate template content from field_definitions JSON
+        Optionally write template to file
+        """
+        pass
+
+    @classmethod
+    def template_to_field_definitions(cls, template_filepath):
+        """
+        Generate field_definitions JSON from tabular template file
+        """
+        pass
+
+```
+
+```python
+# creator/data_templates/views.py
+
+def download_template_version(request, template_version_id):
+    """
+    For specific template version, generate template field definitions file
+    and submission file
+    Zip up above files into archive
+    Return archive content in HTTP response
+    """
+    pass
+
+def download_latest_template(request, data_template_id):
+    """
+    For latest version of a data template, generate template field definitions
+    file and submission file
+    Zip up above files into archive
+    Return archive content in HTTP response
+    """
+
+def download_study_templates(request, study_id):
+    """
+    For all templates assigned to a study, generate template field definitions
+    file and submission files for each template
+    Zip up above files into archive
+    Return archive content in HTTP response
+    """
+    pass
+```
```

Comment:
Maybe a user would want to download all templates assigned to a study that their organization created?